### ISSUE TITLE
Using filter_by and other Typesense search parameters

### ISSUE THREAD

**Issue by furnnl** [typesense/typesense-instantsearch-adapter#17]
## Description

As the documentation says, it should be possible to specify additional search parameters. queryBy works, but filterBy doesn’t.

## Steps to reproduce

I tried the following parameters:

additionalSearchParameters: {
		  queryBy: 'name',
		  filterBy: 'myfilter:0',
		  'filter_by': 'myfilter:0',
		}

## Expected Behavior

Filtered results

## Actual Behavior

Unfiltered results

## Metadata

**Typsense Version**:

**OS**:

**Comment by jasonbosco** [comment]
@furnnl The reason filterBy doesn't work when set in `additionalSearchParameters` is because it's overridden by InstantSearch's refinement widgets internally. So you want to manage default filters through InstantSearch's [`configure` widget](https://www.algolia.com/doc/api-reference/widgets/configure/js/). (Also works similarly in React, Vue, Angular, etc flavors of InstantSearch).

And specifically for filters, you want to use the `facetFilters` key and define the filters in the format described [here](https://www.algolia.com/doc/api-reference/api-parameters/facetFilters/). The adapter will then take care of translating that into the Typesense filter format.

So in your example, you want to do something like this:

```
instantsearch.widgets.configure({ 
  facetFilters: ['myfilter:0']
});
```

Does that help?

**Comment by jasonbosco** [comment]
Closing this for now, but please feel free to re-open if you're still running into issues.

**Comment by shrugs** [comment]
Adding additional info for any react users: 

```tsx
const facetFilters = useMemo(() => [`myfilter:0`], [])

// ...

<Configure facetFilters={facetFilters} />
```

does the trick (making sure to keep the identity of that `facetFilters` array as constant as possible, since it'll trigger re-fetches when it changes, and providing that inline can cause an infinite fetch-render loop)

**Comment by ctorgalson** [comment]
I don't know if it's worth re-opening this but in my tests, adding a correctly-formatted facet filter to `<Configure />` in React adds an a property like `configure: {facetFilters: ['meal:lunch']}` to `searchState`, and adds e.g. `&configure[facetFilters][0]=meal%3Alunch` to the query string, but the actual query results are completely unaffected.

/shrug

**Comment by jasonbosco** [comment]
@ctorgalson Could you post the full `<Configure />` line, showing the `facetFilters` string? 

Could you also copy-as-curl the request that gets made to the Typesense server from the browser and post it?

**Comment by ctorgalson** [comment]
@jasonbosco I will, though I'll have to recreate the issue as I found an alternate solution :)

---

Adding to this thread because I didn't find it on my first ten trips through Algolia's docs: [Virtual Widgets](https://www.algolia.com/doc/guides/building-search-ui/widgets/customize-an-existing-widget/react/#virtual-widgets).

TL;DR: There's a way in React to provide default values without adding a "real" widget, like this:

```jsx
<VirtualRefinementList attribute="categories" defaultRefinement="Cell Phones" />
```

**Comment by reinoldus** [comment]
Has anyone solved it for vue-instantsearch?

If I do: `<ais-configure :filters="'my_filter:true'">` the filter is not picked up, for `<ais-configure :facetFilters="['my_filter:true']">` the filter is picked up, but overwritten once I select another filter

**Comment by jasonbosco** [comment]
@reinoldus Filters need to be specified in the format mentioned [here](https://www.algolia.com/doc/api-reference/api-parameters/facetFilters/). The adapter then converts them to the Typesense filter string format internally. 

> the filter is picked up, but overwritten once I select another filter

This is expected behavior for the configure widget. It's only meant to set a default filter on load. If you need to enforce a filter for all searches, then you want to use the [helper object](https://github.com/algolia/algoliasearch-helper-js#add-a-facet-filter) that's exposed via the [search function](https://www.algolia.com/doc/api-reference/widgets/instantsearch/vue/#widget-param-search-function).

**Comment by krisimmig** [comment]
Hi @jasonbosco , I have tried to use the search helper you mentioned which gets injected into the searchFunction of the as described here. Now I want to add a default filter to all searches based on a language field we have in our typesense index so the user only sees english or german or french results. The field in the index is defined as a facet but when I try to use for example this code:
```js
searchFunction(helper) {
  helper.addFacetRefinement("language", "de").search();
},
```
I get the error that the facet 'language' is not defined. It is mentioned that these facets need to be declared when initializing the helper like so: 
```js
var helper = AlgoliasearchHelper(client, indexName, {
  facets: ['nameOfTheAttribute']
});
```
Is there a way to use facets like that at all using the helper object? I am not sure how I can get to this using the vue version of the adapter and the instansearch-js widget set for vue.

**Comment by jasonbosco** [comment]
@krisimmig I just tried this out myself in Vue and I saw the same error you saw: "Error: X is not defined in the facets attribute of the helper configuration". Didn't find anything useful googling around. I was almost going to give up, but by some trial-and-error serendipity I found a configuration that made that error go away:

```
export default {
  data() {
    return {
      searchClient,
      searchParameters: {
        facets: ['language'],
      },
      searchFunction(helper) {
        helper.addFacetRefinement('language', 'de').search();
      },
    };
  },
};
```

Here's the full diff to get static filters working in my demo Vue app: https://github.com/typesense/typesense-vue-instantsearch-demo/compare/typesense:c4d2400...typesense:7edf1bf

**Comment by krisimmig** [comment]
That´s it, awesome! :) I didn't think to find the facets param in the `ais-configure` widget, but it is there and like that it seems to work as expected. Thanks for your help!

**Comment by pjatx** [comment]
@jasonbosco Thanks for all the tips, this has been driving me insane for days. 

Can you think of a way to conditionally apply the filterBy logic if a given input has a value in it? Use case is being able to filter by an exact record ID should the user want to do that.

**Comment by jasonbosco** [comment]
@pjatx You could do this outside of the context of the adapter, from the Typesense backend. You can turn off typo tolerance and prefix search for just the record ID field, and add it to the query_by fields may be in the beginning of the list. Then if a user searches for an exact record ID, that will be surfaced first.

**Comment by pjatx** [comment]
@jasonbosco thanks that's super helpful - will give it a shot!

**Comment by britisharmy** [comment]
Supported operands, less than is not supported on numeric filter

```
  methods: {
     searchFunction(helper) {
        //helper.addFacetRefinement('deal', 'deal').search();

        helper.addNumericRefinement('price_per_night','>',this.slider_amount).search();

     //helper.addFacetRefinement('price_per_night','>'+this.slider_amount).search();
      },
```

Even the commented code works, but for numeric filter, less than is not supported.

**Comment by britisharmy** [comment]
> @krisimmig I just tried this out myself in Vue and I saw the same error you saw: "Error: X is not defined in the facets attribute of the helper configuration". Didn't find anything useful googling around. I was almost going to give up, but by some trial-and-error serendipity I found a configuration that made that error go away:
> 
> ```
> export default {
>   data() {
>     return {
>       searchClient,
>       searchParameters: {
>         facets: ['language'],
>       },
>       searchFunction(helper) {
>         helper.addFacetRefinement('language', 'de').search();
>       },
>     };
>   },
> };
> ```
> 
> Here's the full diff to get static filters working in my demo Vue app: [typesense/typesense-vue-instantsearch-demo@typesense:c4d2400...typesense:7edf1bf](https://github.com/typesense/typesense-vue-instantsearch-demo/compare/typesense:c4d2400...typesense:7edf1bf)

The searchFunction should be placed inside the functions block in vue js. It still works where its placed but if you have a variable, its changeable once its in the methods block._

**Comment by jasonbosco** [comment]
@britisharmy 

You want to call `addNumericRefinement` in a method chain like this:

https://github.com/typesense/typesense-vue-instantsearch-demo/commit/0b30a5b62fc938196c24658c1ea8727621d750de

Side note: instantsearch only issues `<=` and `>=` filters from its widgets, so those are the only two operators supported by the adapter.

Separately, good call out on moving `searchFunction` to methods. I've updated the demo.

**Comment by pjatx** [comment]
@jasonbosco any way to have a separate search field that only filters on a single attribute within the context of the search adapter? Still wrestling with the same use case above. 

More context: 
- Currently using vue-instantsearch (works great)
- Using it to search through about 50K film records - fields include title, abstract, release date, film ID, etc...
- Want to use main search field (AisSearchBox) to search across all fields, but want to include a separate field that would filter all the results by an exact match on the film id. Using a refinement, but not the best UX. 

Any way this is achievable in my current setup?

**Comment by jasonbosco** [comment]
@pjatx I see, if you really wanted to, you can have multiple search boxes, you'd need multiple instances of InstantSearch as described [here](https://discourse.algolia.com/t/setting-up-instantsearch-with-multiple-search-boxes/11375/4).

`numTypos: 0` will disable typo tolerance for all fields. `numTypos: '0,2,2,2,2,2,2'` is what you need. Could you expand on what you mean by it doesn't work? Did you see an error or did you get results which included typo corrected ones? In any case, could you open a separate issue in the main typesense repo for this?

**Comment by pjatx** [comment]
@jasonbosco thanks, that's definitely helpful. 

Yeah no problem, here's the error I get when I try to pass a string of comma-separated values: 

`Error: 400 - Parameter `num_typos` must be an unsigned integer.`

Would be happy to open a separate issue if helpful!

**Comment by jasonbosco** [comment]
Could you make sure you're running the latest version of Typesense `v0.21.0` - that's the one that has support for per-field `num_typos`. If you're running on Typesense Cloud, if you email contact @ typesense d0t org with your cluster id, we can upgrade it in place for you.

**Comment by pjatx** [comment]
@jasonbosco ah - that's it. I'm still on 0.19.0. Will upgrade and let you know if I have any more issues. Can't thank you enough for the help! Love Typesense : )

**Comment by panoet** [comment]
> 
> 
> @furnnl The reason filterBy doesn't work when set in `additionalSearchParameters` is because it's overridden by InstantSearch's refinement widgets internally. So you want to manage default filters through InstantSearch's [`configure` widget](https://www.algolia.com/doc/api-reference/widgets/configure/js/). (Also works similarly in React, Vue, Angular, etc flavors of InstantSearch).
> 
> And specifically for filters, you want to use the `facetFilters` key and define the filters in the format described [here](https://www.algolia.com/doc/api-reference/api-parameters/facetFilters/). The adapter will then take care of translating that into the Typesense filter format.
> 
> So in your example, you want to do something like this:
> 
> ```
> instantsearch.widgets.configure({ 
>   facetFilters: ['myfilter:0']
> });
> ```
> 
> Does that help?

So, we only can filter by field that has beed declared "facet"? In my case, user want to filter by some value that I didnot make it as facet when creating the collection. So, do I need to recreated the collection and import all the data? I didn't find any "edit schema" menu.

**Comment by jasonbosco** [comment]
> So, we only can filter by field that has beed declared "facet"?

In Typesense Server, you can filter by fields that are not set as facets. However, (unfortunately) InstantSearch.js's widgets makes the assumption that all filter fields are set as facets, since that's how Algolia works. So if you're using InstantSearch with Typesense, you would also have to set any fields you want to filter by as a facet. 

>  So, do I need to recreated the collection and import all the data?

Correct. I'd recommend using [collection aliases](https://typesense.org/docs/0.21.0/api/collection-alias.html) to do this with zero downtime. We have an open item to add support for editing an existing schema: https://github.com/typesense/typesense/issues/96

**Comment by Koniouchine** [comment]
For those trying to configure Angular ~12, here's a working example. 
It also includes a geosearch solution that @jasonbosco helped me solve. 

component.ts
```
// region Typesense Configuration
config = {
  indexName: 'events',
  searchClient
} as InstantSearchConfig;

searchParameters = {
  hitsPerPage: 10,
  q         : '*',
  query_by  : 'title',
  aroundLatLng: '45.406431, -78.6848739',
  aroundRadius: 1000
}
// endregion
```

component.html
```
<ais-instantsearch [config]="config" (change)="setHits($event)">
  <ais-configure
    [searchParameters]="searchParameters"
  ></ais-configure>

  <ais-search-box [searchAsYouType]="false"></ais-search-box>
  <ais-geo-search></ais-geo-search>

  <ais-hits>
    <ng-template #searchResults let-hits="hits" let-results="results">
      <div *ngFor="let hit of hits">
        {{hit.title}}
      </div>
    </ng-template>
  </ais-hits>
</ais-instantsearch>
```

**Comment by Koniouchine** [comment]
Further insight for filtering with numbers within Angular: 
As previously mentioned, when doing a numeric search, only ">= and <=" are supported. 
However you must also be use them with the 'numericFilters' field instead of 'facetFilters'. 

```
// assign to <ais-config> widget
  searchParameters: any = {
    hitsPerPage: 10,
    numericFilters: ['date.startDate >= 1647068400'],
    q         : '*',
    query_by  : 'title',
    aroundLatLng: '44.406431, -79.6848739',
    aroundRadius: 14000 // in meters
  }
```

**Comment by jasonbosco** [comment]
Quick update: I just added the ability for the adapter to handle the `filters` prop in the `Configure` widget in addition to `facetFilters` and `numericFilters` in v`2.4.1-3`. [Here's](https://github.com/typesense/typesense-instantsearch-adapter#configure) documentation for the format to use with `filters`.

**Comment by uncvrd** [comment]
@jasonbosco came here via a Slack Typesense thread. I just gave `numericFilters` a shot on `2.4.1-3` and I'm receiving the following error:

<img width="732" alt="image" src="https://user-images.githubusercontent.com/25097099/169490283-3b24dd78-d2e2-46cf-b22f-d47fe6eaf59e.png">

Per typsense docs, I think I'm implementing it correctly by following Algolia's syntax, right?:

```
<Configure 
    numericFilters={[
        `teamId = 1`
    ]}
/>
```

However, with this version, I can at least get away with using `filters` as that seems to work in `2.4.1-3`:

```
<Configure 
    filters='teamId:= 3'
/>
```

Am I missing something obvious in the numericFilters implementation?

**Comment by jasonbosco** [comment]
@uncvrd Thank you for catching that and thank you for PR #114. I've published this in `v2.4.1-4`.

**Comment by weiklr** [comment]
this was really helpful in figuring out how to use typesense search parameters with `configure` i managed to get facet refinement working as follows, (using js version of instantsearch):

`
 facetWidget = instantsearch.widgets.configure({
            facets: ['your facet column']
   })
searchInstance = instantsearch({indexName, typesenseSearchClient})
searchInstance.addWidget(facetWidget)
searchInstance.start()
searchInstance.helper.addFacetRefinement('facet column name', 'facet value'))
`

**Comment by braincomb** [comment]
Hi @jasonbosco ,

Is there a way to do this with negation? For example we want to show all items EXCEPT those that have `someAttr:true`. How would we do this `:facetFilters`?

Thanks.

> @reinoldus Filters need to be specified in the format mentioned [here](https://www.algolia.com/doc/api-reference/api-parameters/facetFilters/). The adapter then converts them to the Typesense filter string format internally.
> 
> > the filter is picked up, but overwritten once I select another filter
> 
> This is expected behavior for the configure widget. It's only meant to set a default filter on load. If you need to enforce a filter for all searches, then you want to use the [helper object](https://github.com/algolia/algoliasearch-helper-js#add-a-facet-filter) that's exposed via the [search function](https://www.algolia.com/doc/api-reference/widgets/instantsearch/vue/#widget-param-search-function).

**Comment by jasonbosco** [comment]
@braincomb Does `addFacetExclusion` help: https://instantsearchjs.netlify.app/algoliasearch-helper-js/reference#AlgoliaSearchHelper#addFacetExclusion

**Comment by braincomb** [comment]
> @braincomb Does `addFacetExclusion` help: https://instantsearchjs.netlify.app/algoliasearch-helper-js/reference#AlgoliaSearchHelper#addFacetExclusion

Thanks @jasonbosco, that worked perfectly.

For anyone looking for the same solution, here's what I did for `vue-instantsearch` to do an exclusion filter:

```vue
<ais-instant-search index-name="index" :search-client="searchClient" :search-function="searchHelper">
```

```vue
<ais-configure :facets="['facetToExclude']"></ais-configure>
```

```ts
const searchHelper= (helper: any) => {
  helper.addFacetExclusion("facetToExclude", "value").search();
};
```
